lib/Service/IMipService.php

```diff
@@ -0,0 +1,169 @@
+use OCA\Mail\Account;
+use OCA\Mail\Db\Mailbox;
+use OCA\Mail\Db\MailboxMapper;
+use OCA\Mail\Db\Message;
+use OCA\Mail\Db\MessageMapper;
+use OCA\Mail\Exception\ServiceException;
+use OCA\Mail\Model\IMAPMessage;
+use OCP\AppFramework\Db\DoesNotExistException;
+use OCP\Calendar\IManager;
+use Psr\Log\LoggerInterface;
+
+class IMipService {
+	private AccountService $accountService;
+	private IManager $calendarManager;
+	private LoggerInterface $logger;
+	private MailboxMapper $mailboxMapper;
+	private MailManager $mailManager;
+	private MessageMapper $messageMapper;
+
+	public function __construct(
+		AccountService $accountService,
+		IManager $manager,
+		LoggerInterface $logger,
+		MailboxMapper $mailboxMapper,
+		MailManager $mailManager,
+		MessageMapper $messageMapper
+	) {
+		$this->accountService = $accountService;
+		$this->calendarManager = $manager;
+		$this->logger = $logger;
+		$this->mailboxMapper = $mailboxMapper;
+		$this->mailManager = $mailManager;
+		$this->messageMapper = $messageMapper;
+	}
+
+	public function process(): void {
+		$messages = $this->messageMapper->findIMipMessagesAscending();
+		if (empty($messages)) {
+			$this->logger->info('No iMIP messages to process.');
+			return;
+		}
+
+		// Collect all mailboxes in memory
+		// possible perf improvement - make this one IN query
+		// and JOIN with accounts table
+		// although this might not make much of a difference
+		// since there are very few messages to process
+		$mailboxIds = array_unique(array_map(function (Message $message) {
+			return $message->getMailboxId();
+		}, $messages));
+
+		$mailboxes = array_map(function (int $mailboxId) {
+			try {
+				return $this->mailboxMapper->findById($mailboxId);
+			} catch (DoesNotExistException | ServiceException $e) {
+				return null;
+			}
+		}, $mailboxIds);
+
+		// Collect all accounts in memory
+		$accountIds = array_unique(array_map(function (Mailbox $mailbox) {
+			return $mailbox->getAccountId();
+		}, $mailboxes));
+
+		$accounts = array_combine($accountIds, array_map(function (int $accountId) {
+			try {
+				return $this->accountService->findById($accountId);
+			} catch (DoesNotExistException $e) {
+				return null;
+			}
+		}, $accountIds));
+
+		/** @var Mailbox $mailbox */
+		foreach ($mailboxes as $mailbox) {
+			/** @var Account $account */
+			$account = $accounts[$mailbox->getAccountId()];
+			$filteredMessages = array_filter($messages, function ($message) use ($mailbox) {
+				return $message->getMailboxId() === $mailbox->getId();
+			});
+
+			if (empty($filteredMessages)) {
+				continue;
+			}
+
+			// Check for accounts or mailboxes that no longer exist,
+			// no processing for drafts, sent items, junk or trash
+			if ($account === null
+				|| $mailbox->getSpecialUse() === '["sent"]'
+				|| $mailbox->getSpecialUse() === '["archive"]'
+				|| $mailbox->getSpecialUse() === '["junk"]'
+				|| $mailbox->getSpecialUse() === '["drafts"]'
+				|| $mailbox->getSpecialUse() === '["trash"]'
+			) {
```

Comment:
the column is json encoded and if you have multiple uses on one mailbox your change would fail

HOWEVER

we don't use the IMAP special use for the sent/trash/drafts preference. we use account.*_mailbox_id. on IMAP there can be n mailboxes with the same use. There is no guaranteed correlation between the imap flags and our preferences other than we try to detect ours from what is found on imap.

You can see `\OCA\Mail\Service\Classification\ImportanceClassifier::getOutgoingMailboxes` for reference

Reply:
kk, thanks for the info

Reply:
Chnaged it to the mailbox id methods, plus one use case for archive.